chore: design 0 (#2)
fission-codes/stack: design/api.md

  /**
   * Trigger an email flow to register a user
   * Signs a Register UCAN with an Identifier
   */
  register(credentials: RegisterCredentials): Promise<
    AuthResponse<{
      account: Account
    }>
  >

  /**
   * Used to exchange a code from an email link for a session
   */
  exchangeCodeForSession(code: string): Promise<
    AuthResponse<{
      session: Session
      account: Account
    }>
  >

  /**
   * Requests an Identifier Delegation and IF NEEDED Server Delegations
   *
   * Should we control Session TTL with Identifier Delegation exp and/or Server Delegation exp?
   */
  login(credentials: Credentials): Promise<
    AuthResponse<{
      session: Session
      account: Account
    }>
  >

  /**
   * Delete Identifier and server Delegations
   *
   * Should we delete all agent data as well?
   */
  logout(session: Session): Promise<void>

  /**
   * Subscribe to auth state changes
   */
  onStateChange(
    cb: (
      state: 'Login' | 'Logout' | 'Register' | 'Expired',
      session: null | Session
    ) => void
  ): UnsubscribeFn
}
```

### Notes

- Should keep tabs in sync when in browser
- Revocation api ?

### Identifier Providers

> Providers SHOULD:

Reviewer: Is this an RFC2119 `SHOULD`? If so, let's put the disclaimer at the top.

- Depend only on UCAN (did:key, signer) and optionally on a [channel](#channels). They can have other dependencies but they should be environment specific. They SHOULD NOT depend on an [Agent](#agent).
- Be able to receive a DID and a set of capabilities and return a powerbox style UCAN with those capabilities.
- Be able to send a root delegation over a [channel](#channels) to another device or origin. (see [Delegated](#types))

A powerbox style delegation to an Agent should be treated as a Session.

#### Types

- Local: Only exists on device normally sandboxed to an origin (webcrypto, wallet, metamask snap and local passkeys)

> - Synced: Exists on device and is synced with a server (synced passkey)

Reviewer: Where a synced passkey is managed by the browser and has the same DID across devices, or could it run into the problem where there's multiple DIDs for the same user on the same browser that you alluded to the other day?

Reviewer: Yep makes sense. Registering a new passkey is possible two ways, and you're right that we should have APIs the email one:

- Device linking: use one of your client-side UCANs to delegate permissions
- Email recovery: we send you a code, and you tell us which DID to delegate to (similar to Bluesky and other systems, could be extended with OAuth logins etc down the line, though bleh rooting security in OAuth)

Reviewer: 💯

Author: something like metamask could also be considered Synced if we assume the User manually syncs the mnemonic, i just dont think we should... That why metamask is local in the text, and we should either do Local/Delegated or Local/Remote linking.

- Remote: The identifier lives in another device or origin and it needs a [channel](#channels) (websocket, MessageChannel) to request capabilities and receive UCANs. An example could be a personal lobby or a mobile app (wallet).
- Delegated: The identifier has a wildcard delegation from a Remote Identifier Provider. Does NOT have the root keypair just a root delegation. This is useful for Phase 0 device link with QR Code. (same as local but with a wildcard delegation)

> ### Fission Client
> 
> Client SHOULD only handle HTTP logic, it should receive UCAN delegations, set bearer tokens according to [Spec](https://github.com/ucan-wg/ucan-http-bearer-token) and handle HTTP requests.
> 
> In the future this SHOULD be a generic UCAN-RPC client that can be used for any UCAN-RPC protocol, given a set of Capabilities.

Reviewer: Sine we're RFC2119ing this, I'd suggest reframing as...

```suggestion
### Fission Client

The Client MUST support HTTP requests, receive UCAN delegations, set bearer tokens according to [Spec](https://github.com/ucan-wg/ucan-http-bearer-token).

In the future SHOULD be a generic UCAN-RPC client that MAY be used for any UCAN-RPC protocol, given a set of Capabilities.
```

...unless you wanted to make all interfaces optional.

Author: > ...unless you wanted to make all interfaces optional.

what do you mean ?

the text syntax is still very informal but i will RFC2119 it more before merging
